src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -106,17 +109,26 @@ public CandlepinQuery<Pool> listByOwner(Owner owner) {
      * @return pools owned by the given Owner.
      */
     @Transactional
-    public CandlepinQuery<Pool> listByOwner(Owner owner, Date activeOn) {
-        DetachedCriteria criteria = DetachedCriteria.forClass(Pool.class)
-            .add(Restrictions.eq("owner", owner))
-            .add(Restrictions.eq("activeSubscription", Boolean.TRUE));
+    public List<Pool> listByOwner(Owner owner, Date activeOn) {
+        CriteriaBuilder cb = this.getEntityManager().getCriteriaBuilder();
+        CriteriaQuery<Pool> query = cb.createQuery(this.entityType());
+        Root<Pool> root = query.from(this.entityType());
+        query.select(root);
+
+        List<Predicate> predicates = new ArrayList<>();
+        predicates.add(cb.equal(root.get(Pool_.owner), owner));
+        predicates.add(cb.equal(root.get(Pool_.activeSubscription), true));
 
         if (activeOn != null) {
-            criteria.add(Restrictions.le("startDate", activeOn))
-                .add(Restrictions.ge("endDate", activeOn));
+            predicates.add(cb.lessThanOrEqualTo(root.get(Pool_.startDate), activeOn));
+            predicates.add(cb.greaterThanOrEqualTo(root.get(Pool_.endDate), activeOn));
         }
 
-        return this.cpQueryFactory.buildQuery(this.currentSession(), criteria);
+        query.where(predicates.toArray(new Predicate[predicates.size()]));
```

Comment:
It's so crazy that the JPA standard doesn't just accept a list straight away and instead forces this double allocation and copy nonsense on us.

The Predicate interface doesn't look to be too complex, and we could probably build a handy version of that which is mutable or at least has a builder so we could do this without incurring the pain of that second array allocation and copy.

Reply:
I know right! Having to move over the Predicates into a new array is silly haha

I like the idea of having our own version or some sort of builder that can be used throughout the code base to avoid this issue! I can get a Jira logged for this

Reply:
https://issues.redhat.com/browse/CANDLEPIN-803